### PR TITLE
[mlir] [dataflow] unify semantics of program point

### PR DESCRIPTION
**Pull request by cxy-1993**
The concept of a 'program point' in the original data flow framework is ambiguous. It can refer to either an operation or a block itself. This representation has different interpretations in forward and backward data-flow analysis. In forward data-flow analysis, the program point of an operation represents the state after the operation, while in backward data flow analysis, it represents the state before the operation. When using forward or backward data-flow analysis, it is crucial to carefully handle this distinction to ensure correctness.

This patch refactors the definition of program point, unifying the interpretation of program points in both forward and backward data-flow analysis.

How to integrate this patch?

For dense forward data-flow analysis and other analysis (except dense backward data-flow analysis), the program point corresponding to the original operation can be obtained by `getProgramPointAfter(op)`, and the program point corresponding to the original block can be obtained by `getProgramPointBefore(block)`.

For dense backward data-flow analysis, the program point corresponding to the original operation can be obtained by `getProgramPointBefore(op)`, and the program point corresponding to the original block can be obtained by `getProgramPointAfter(block)`.

NOTE: If you need to get the lattice of other data-flow analyses in dense backward data-flow analysis, you should still use the dense forward data-flow approach. For example, to get the Executable state of a block in dense backward data-flow analysis and add the dependency of the current operation, you should write:

``getOrCreateFor<Executable>(getProgramPointBefore(op), getProgramPointBefore(block))``

In case above, we use getProgramPointBefore(op) because the analysis we rely on is dense backward data-flow, and we use getProgramPointBefore(block) because the lattice we query is the result of a non-dense backward data flow computation.

related dsscussion: https://discourse.llvm.org/t/rfc-unify-the-semantics-of-program-points/80671/8
corresponding PSA: https://discourse.llvm.org/t/psa-program-point-semantics-change/81479

### CODE REVIEW

**Review by tblah — APPROVED** [review]
Flang changes LGTM. Please wait for somebody else to approve MLIR changes.

**Comment by cxy-1993** [comment]
@Mogball  Could you please help review this patch, Thanks!

**Comment by cxy-1993** [comment]
ping @Mogball @ftynse @stellaraccident

**Review by stellaraccident — COMMENTED** [review]
This looks ok to me but I'm hardly authoritative on this code. I think @Mogball is probably still the best reviewer and we should give a few more days before finding an alternate. I'm also going to patch this in to a downstream and make sure it doesn't break any tests.

**Comment by cxy-1993** [comment]
ping @Mogball

**Review by Mogball — APPROVED** [review]

File: mlir/lib/Analysis/DataFlow/SparseAnalysis.cpp [hunk 1]
```diff
@@ -542,7 +554,8 @@ void AbstractSparseBackwardDataFlowAnalysis::visitRegionSuccessors(
     MutableArrayRef<OpOperand> opoperands = operandsToOpOperands(operands);
     ValueRange inputs = successor.getSuccessorInputs();
     for (auto [operand, input] : llvm::zip(opoperands, inputs)) {
-      meet(getLatticeElement(operand.get()), *getLatticeElementFor(op, input));
+      meet(getLatticeElement(operand.get()),
+           *getLatticeElementFor(getProgramPointAfter(op), input));
```
Comment (matthias-springer) [inline mlir/lib/Analysis/DataFlow/SparseAnalysis.cpp]: @cxy-1993 I looked at this code today. Is this correct? `input` could be a block argument of a region. Is it correct to always use `getProgramPointAfter(op)` or should this distinguish between block argument / op result?

**Pull request merged**